rust/src/delta_datafusion.rs

```diff
@@ -673,6 +748,118 @@ impl DeltaDataChecker {
+            .map_err(|_| DataFusionError::Internal("Unable to decode DeltaScan".to_string()))?;
+        let delta_scan = DeltaScan {
+            url: msg.url.clone(),
+            parquet_scan: (*inputs)[0].clone(),
+        };
+        Ok(Arc::new(delta_scan))
+    }
+
+    fn try_encode(
+        &self,
+        node: Arc<dyn ExecutionPlan>,
+        buf: &mut Vec<u8>,
+    ) -> Result<(), DataFusionError> {
+        let delta_scan = node
+            .as_any()
+            .downcast_ref::<DeltaScan>()
+            .ok_or(DataFusionError::Internal("Not a delta scan!".to_string()))?;
+        let msg = DeltaTableProto {
+            url: delta_scan.url.clone(),
+        };
+        msg.encode(buf)
+            .map_err(|_| DataFusionError::Internal("Unable to encode delta scan!".to_string()))?;
+        Ok(())
+    }
+}
+
+/// Does serde on DeltaTables
+#[derive(Debug)]
+pub struct DeltaLogicalCodec {}
+
+#[async_trait]
+impl LogicalExtensionCodec for DeltaLogicalCodec {
+    fn try_decode(
+        &self,
+        _buf: &[u8],
+        _inputs: &[LogicalPlan],
+        _ctx: &SessionContext,
+    ) -> Result<Extension, DataFusionError> {
+        todo!("DeltaLogicalCodec")
+    }
+
+    fn try_encode(&self, _node: &Extension, _buf: &mut Vec<u8>) -> Result<(), DataFusionError> {
+        todo!("DeltaLogicalCodec")
+    }
+
+    async fn try_decode_table_provider(
+        &self,
+        buf: &[u8],
+        _schema: SchemaRef,
+        ctx: &SessionContext,
+    ) -> Result<Arc<dyn TableProvider>, DataFusionError> {
+        let msg = DeltaTableProto::decode(buf)
+            .map_err(|_| DataFusionError::Internal("Error encoding delta table".to_string()))?;
+        let factory = ctx
+            .state
+            .read()
+            .runtime_env
+            .table_factories
+            .get("deltatable")
+            .ok_or_else(|| DataFusionError::Plan(format!("Unable to find deltatable factory",)))?
+            .clone();
+        let provider = (*factory).create(msg.url.as_str()).await?;
+        Ok(provider)
+    }
+
+    fn try_encode_table_provider(
+        &self,
+        node: Arc<dyn TableProvider>,
+        buf: &mut Vec<u8>,
+    ) -> Result<(), DataFusionError> {
+        let table = node.as_ref().as_any().downcast_ref::<DeltaTable>().ok_or(
+            DataFusionError::Internal("Can't encode non-delta tables".to_string()),
+        )?;
+        let msg = DeltaTableProto {
+            url: table.storage.root_uri(),
+        };
+        msg.encode(buf)
+            .map_err(|_| DataFusionError::Internal("Error encoding delta table".to_string()))
+    }
+}
+
+/// Responsible for creating deltatables
+pub struct DeltaTableFactory {}
+
+#[async_trait]
+impl TableProviderFactory for DeltaTableFactory {
+    async fn create(&self, url: &str) -> datafusion::error::Result<Arc<dyn TableProvider>> {
+        let provider = open_table(url).await.unwrap();
+        Ok(Arc::new(provider))
```

Comment:
This is perhaps a place where `storage_options` will get lost, if we care about that. If we only want to support passing credentials via env variables for systems that use this (just distributed, right?), that might be sensible.

Reply:
> we only want to support passing credentials via env variables for systems that use this (just distributed, right?), that might be sensible.

For now, that is all I am attempting to do with this PR. I can possibly see someone wanting to file another PR to extend it in the future, but if they are passing `storage_options`, I think that means they are registering tables in Rust (not SQL), and therefor already are writing a fork of Ballista, in which case they can write their own serde?

It's a worthwhile thought experiment, but I wouldn't recommend coding for it quite yet...

Reply:
That's fair. I just wanted to make sure you were aware of that limitation.